File: netlify/functions/health-check.py
import json
import os
import requests

def handler(event, context):
    """Check health status of all integrations"""
    
    headers = {
        "Content-Type": "application/json",
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Headers": "Content-Type, Authorization",
        "Access-Control-Allow-Methods": "GET, OPTIONS"
    }
    
    if event.get("httpMethod") == "OPTIONS":
        return {"statusCode": 200, "headers": headers, "body": ""}
    
    try:
        # Get user's access token if authenticated
        auth_header = event.get("headers", {}).get("authorization") or event.get("headers", {}).get("Authorization")
        google_access_token = None
        if auth_header and auth_header.startswith("Bearer "):
            google_access_token = auth_header[7:]
        
        health = {
            "meta": check_meta_health(),
            "google_oauth": check_google_oauth(google_access_token),
            "google_sheets": check_google_sheets(google_access_token),
            "google_drive": check_google_drive(google_access_token),
            "config": get_config_status()
        }
        
        return {
            "statusCode": 200,
            "headers": headers,
            "body": json.dumps(health)
        }
        
    except Exception as e:
        return {
            "statusCode": 500,
            "headers": headers,
            "body": json.dumps({"error": str(e)})
        }

def check_meta_health():
    """Check Meta Ads API connection"""
    meta_token = os.environ.get("META_ACCESS_TOKEN")
    ad_account_id = os.environ.get("META_AD_ACCOUNT_ID")
    
    if not meta_token or not ad_account_id:
        return {
            "status": "not_configured",
            "message": "Meta credentials not configured",
            "configured": False
        }
    
    try:
        # Test API call
        url = f"https://graph.facebook.com/v21.0/{ad_account_id}"
        response = requests.get(url, params={
            "access_token": meta_token,
            "fields": "name,account_status"
        }, timeout=10)
        
        if response.status_code == 200:
            data = response.json()
            return {
                "status": "connected",
                "message": f"Connected to {data.get('name', 'Unknown')}",
                "configured": True,
                "account_name": data.get("name"),
                "account_status": data.get("account_status")
            }
        else:
            error_data = response.json()
            return {
                "status": "error",
                "message": error_data.get("error", {}).get("message", "API error"),
                "configured": True
            }
    except Exception as e:
        return {
            "status": "error",
            "message": str(e),
            "configured": True
        }

def check_google_oauth(access_token):
    """Check Google OAuth status"""
    if not access_token:
        client_id = os.environ.get("GOOGLE_CLIENT_ID")
        if not client_id:
            return {
                "status": "not_configured",
                "message": "Google OAuth not configured",
                "configured": False
            }
        return {
            "status": "not_authenticated",
            "message": "OAuth configured but user not logged in",
            "configured": True
        }
    
    try:
        # Verify token with userinfo endpoint
        response = requests.get(
            "https://www.googleapis.com/oauth2/v2/userinfo",
            headers={"Authorization": f"Bearer {access_token}"},
            timeout=10
        )
        
        if response.status_code == 200:
            user_info = response.json()
            return {
                "status": "connected",
                "message": f"Logged in as {user_info.get('email')}",
                "configured": True,
                "user_email": user_info.get("email"),
                "user_name": user_info.get("name")
            }
        else:
            return {
                "status": "error",
                "message": "Invalid or expired token",
                "configured": True
            }
    except Exception as e:
        return {
            "status": "error",
            "message": str(e),
            "configured": True
        }

def check_google_sheets(access_token):
    """Check Google Sheets API access"""
    if not access_token:
        return {
            "status": "not_authenticated",
            "message": "Login with Google to access Sheets",
            "configured": False
        }
    
    spreadsheet_id = os.environ.get("GOOGLE_SPREADSHEET_ID")
    if not spreadsheet_id:
        return {
            "status": "warning",
            "message": "No spreadsheet ID configured (using mock data)",
            "configured": False
        }
    
    try:
        # Test read access
        url = f"https://sheets.googleapis.com/v4/spreadsheets/{spreadsheet_id}"
        response = requests.get(
            url,
            headers={"Authorization": f"Bearer {access_token}"},
            params={"fields": "properties(title)"},
            timeout=10
        )
        
        if response.status_code == 200:
            data = response.json()
            title = data.get("properties", {}).get("title", "Unknown")
            return {
                "status": "connected",
                "message": f"Connected to: {title}",
                "configured": True,
                "spreadsheet_title": title
            }
        else:
            return {
                "status": "error",
                "message": "Cannot access spreadsheet",
                "configured": True
            }
    except Exception as e:
        return {
            "status": "error",
            "message": str(e),
            "configured": True
        }

def check_google_drive(access_token):
    """Check Google Drive API access"""
    if not access_token:
        return {
            "status": "not_authenticated",
            "message": "Login with Google to access Drive",
            "configured": False
        }
    
    try:
        # Test Drive access
        url = "https://www.googleapis.com/drive/v3/about"
        response = requests.get(
            url,
            headers={"Authorization": f"Bearer {access_token}"},
            params={"fields": "user,storageQuota"},
            timeout=10
        )
        
        if response.status_code == 200:
            data = response.json()
            user = data.get("user", {})
            quota = data.get("storageQuota", {})
            
            used_gb = int(quota.get("usage", 0)) / (1024**3)
            limit_gb = int(quota.get("limit", 0)) / (1024**3)
            
            return {
                "status": "connected",
                "message": f"Connected as {user.get('emailAddress')}",
                "configured": True,
                "storage_used_gb": round(used_gb, 2),
                "storage_limit_gb": round(limit_gb, 2)
            }
        else:
            return {
                "status": "error",
                "message": "Cannot access Drive",
                "configured": True
            }
    except Exception as e:
        return {
            "status": "error",
            "message": str(e),
            "configured": True
        }

def get_config_status():
    """Get current configuration status"""
    return {
        "meta": {
            "access_token_set": bool(os.environ.get("META_ACCESS_TOKEN")),
            "ad_account_id": os.environ.get("META_AD_ACCOUNT_ID", "Not set")
        },
        "google_oauth": {
            "client_id_set": bool(os.environ.get("GOOGLE_CLIENT_ID")),
            "client_secret_set": bool(os.environ.get("GOOGLE_CLIENT_SECRET")),
            "redirect_uri": os.environ.get("GOOGLE_REDIRECT_URI", "Not set")
        },
        "google_sheets": {
            "spreadsheet_id": os.environ.get("GOOGLE_SPREADSHEET_ID", "Not set")
        }
    }

File: netlify/functions/health-check.js
const fetch = require('node-fetch');

exports.handler = async (event, context) => {
  const headers = {
    'Content-Type': 'application/json',
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Headers': 'Content-Type, Authorization',
    'Access-Control-Allow-Methods': 'GET, OPTIONS'
  };

  if (event.httpMethod === 'OPTIONS') {
    return { statusCode: 200, headers, body: '' };
  }

  try {
    const authHeader = event.headers.authorization || event.headers.Authorization;
    const googleAccessToken = authHeader?.startsWith('Bearer ') ? authHeader.slice(7) : null;

    const health = {
      meta: await checkMetaHealth(),
      google_oauth: await checkGoogleOAuth(googleAccessToken),
      google_sheets: await checkGoogleSheets(googleAccessToken),
      google_drive: await checkGoogleDrive(googleAccessToken),
      config: getConfigStatus()
    };

    return {
      statusCode: 200,
      headers,
      body: JSON.stringify(health)
    };

  } catch (error) {
    return {
      statusCode: 500,
      headers,
      body: JSON.stringify({ error: error.message })
    };
  }
};

async function checkMetaHealth() {
  const metaToken = process.env.META_ACCESS_TOKEN;
  const adAccountId = process.env.META_AD_ACCOUNT_ID;

  if (!metaToken || !adAccountId) {
    return {
      status: 'not_configured',
      message: 'Meta credentials not configured',
      configured: false
    };
  }

  try {
    const url = `https://graph.facebook.com/v21.0/${adAccountId}`;
    const params = new URLSearchParams({
      access_token: metaToken,
      fields: 'name,account_status'
    });

    const response = await fetch(`${url}?${params}`);
    if (response.ok) {
      const data = await response.json();
      return {
        status: 'connected',
        message: `Connected to ${data.name || 'Unknown'}`,
        configured: true,
        account_name: data.name,
        account_status: data.account_status
      };
    } else {
      const error = await response.json();
      return {
        status: 'error',
        message: error.error?.message || 'API error',
        configured: true
      };
    }
  } catch (error) {
    return {
      status: 'error',
      message: error.message,
      configured: true
    };
  }
}

async function checkGoogleOAuth(accessToken) {
  if (!accessToken) {
    const clientId = process.env.GOOGLE_CLIENT_ID;
    if (!clientId) {
      return {
        status: 'not_configured',
        message: 'Google OAuth not configured',
        configured: false
      };
    }
    return {
      status: 'not_authenticated',
      message: 'OAuth configured but user not logged in',
      configured: true
    };
  }

  try {
    const response = await fetch('https://www.googleapis.com/oauth2/v2/userinfo', {
      headers: { 'Authorization': `Bearer ${accessToken}` }
    });

    if (response.ok) {
      const userInfo = await response.json();
      return {
        status: 'connected',
        message: `Logged in as ${userInfo.email}`,
        configured: true,
        user_email: userInfo.email,
        user_name: userInfo.name
      };
    } else {
      return {
        status: 'error',
        message: 'Invalid or expired token',
        configured: true
      };
    }
  } catch (error) {
    return {
      status: 'error',
      message: error.message,
      configured: true
    };
  }
}

async function checkGoogleSheets(accessToken) {
  if (!accessToken) {
    return {
      status: 'not_authenticated',
      message: 'Login with Google to access Sheets',
      configured: false
    };
  }

  const spreadsheetId = process.env.GOOGLE_SPREADSHEET_ID;
  if (!spreadsheetId) {
    return {
      status: 'warning',
      message: 'No spreadsheet ID configured (using mock data)',
      configured: false
    };
  }

  try {
    const url = `https://sheets.googleapis.com/v4/spreadsheets/${spreadsheetId}?fields=properties(title)`;
    const response = await fetch(url, {
      headers: { 'Authorization': `Bearer ${accessToken}` }
    });

    if (response.ok) {
      const data = await response.json();
      const title = data.properties?.title || 'Unknown';
      return {
        status: 'connected',
        message: `Connected to: ${title}`,
        configured: true,
        spreadsheet_title: title
      };
    } else {
      return {
        status: 'error',
        message: 'Cannot access spreadsheet',
        configured: true
      };
    }
  } catch (error) {
    return {
      status: 'error',
      message: error.message,
      configured: true
    };
  }
}

async function checkGoogleDrive(accessToken) {
  if (!accessToken) {
    return {
      status: 'not_authenticated',
      message: 'Login with Google to access Drive',
      configured: false
    };
  }

  try {
    const url = 'https://www.googleapis.com/drive/v3/about?fields=user,storageQuota';
    const response = await fetch(url, {
      headers: { 'Authorization': `Bearer ${accessToken}` }
    });

    if (response.ok) {
      const data = await response.json();
      const user = data.user || {};
      const quota = data.storageQuota || {};

      const usedGb = parseFloat(quota.usage || 0) / (1024**3);
      const limitGb = parseFloat(quota.limit || 0) / (1024**3);

      return {
        status: 'connected',
        message: `Connected as ${user.emailAddress}`,
        configured: true,
        storage_used_gb: Math.round(usedGb * 100) / 100,
        storage_limit_gb: Math.round(limitGb * 100) / 100
      };
    } else {
      return {
        status: 'error',
        message: 'Cannot access Drive',
        configured: true
      };
    }
  } catch (error) {
    return {
      status: 'error',
      message: error.message,
      configured: true
    };
  }
}

function getConfigStatus() {
  return {
    meta: {
      access_token_set: !!process.env.META_ACCESS_TOKEN,
      ad_account_id: process.env.META_AD_ACCOUNT_ID || 'Not set'
    },
    google_oauth: {
      client_id_set: !!process.env.GOOGLE_CLIENT_ID,
      client_secret_set: !!process.env.GOOGLE_CLIENT_SECRET,
      redirect_uri: process.env.GOOGLE_REDIRECT_URI || 'Not set'
    },
    google_sheets: {
      spreadsheet_id: process.env.GOOGLE_SPREADSHEET_ID || 'Not set'
    }
  };
}
